### PR TITLE
initial commit - countdown clock for giving tues

### PR DESCRIPTION
**Pull request by wmui51**
# READY FOR REVIEW

# Summary
- Created a Countdown Clock component for the Giving Tuesday campaign. The `Countdown` component contains `CountdownPie` components, each pie represents a unit of time for the countdown and styling is done via scss and a dynamic inline `<style>` tag. A `UseCountdown` hook is also included, it does the counting down from a specific date. This is setup to be dropped in as is via Storyblok, as counting down to 11/28 is hard coded and per the design request, only Hours, Minutes, and Seconds are shown.

This is for MVP as to make sure that this meets the needs of the stakeholders. There will be a subsequent PR to make this more reusable and make the countdown date, which pies to show, pie width, and colors all configurable via Storyblok.

# Review By October 26, 2023

- This component will be used on the Giving Tuesday page (currently in draft) https://giving.stanford.edu/giving-tuesday/. We would like to begin testing the page in the first week of November.

# Criticality - 7
- The team is expecting this component to be available and ties into donor-facing messaging.

# Review Tasks

## Setup tasks and/or behavior to test

1. Check out this branch
2. Navigate to...
3. Verify...

## Front End Validation
- [ ] Is the markup using the appropriate semantic tags and passes HTML validation?
- [ ] Cross-browser testing has been performed?
- [ ] Automated accessibility scans performed?
- [ ] Manual accessibility tests performed?
- [ ] Design is approved by @ user?

### Code
- [ ] Are the naming conventions following our standards?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided? eg (unit, behat, or codeception)

### Code security
- [ ] Are all [forms properly sanitized](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

## General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

# Affected Projects or Products
- Does this PR impact any particular projects, products, or modules?

# Associated Issues and/or People
- JIRA ticket(s)
- Other PRs
- Any other contextual information that might be helpful (e.g., description of a bug that this PR fixes, new functionality that it adds, etc.)
- Anyone who should be notified? (`@mention` them here)

# Resources
- [AMP Tool](https://stanford.levelaccess.net/index.php)
- [Accessibility Manual Test Script](https://docs.google.com/document/d/1ZXJ9RIUNXsS674ow9j3qJ2g1OAkCjmqMXl0Gs8XHEPQ/edit?usp=sharing)
- [HTML Validator](https://validator.w3.org/)
- [Browserstack](https://live.browserstack.com/dashboard) and link to [Browserstack Credentials](https://asconfluence.stanford.edu/confluence/display/SWS/External+Account+Credentials)

### CODE REVIEW

File: package.json
Comment (sherakama) [inline package.json:8]: @wmui51,

This is a new CSS convention for this website and it requires a [large bundle](https://bundlephobia.com/package/@emotion/react@11.11.1). Is it possible to avoid adding this package and using what exists today or are you blocked by something?

File: src/components/endowed-positions/EndowedPositionsHeader.js
Comment (sherakama) [inline src/components/endowed-positions/EndowedPositionsHeader.js:1]: If you need this file to be `jsx` just rename the file extension to be so. Gatsby supports JSX already.

File: package.json
Comment (sherakama) [reply package.json:8]: I have thoughts but perhaps @yvonnetangsu and/or @rebeccahongsf have better thoughts first?

Comment (yvonnetangsu) [reply package.json:8]: @wmui51 @sherakama I haven't looked into this PR, but could we use `dcnb` (for dynamically chaining class names)?

Comment (yvonnetangsu) [reply package.json:8]: @wmui51 Hmm Thanks for the preview - that really helps. I'm wondering if this UX works well for this situation - seems a bit tough to make it work well on mobile also. Personally I feel like the donut chart design is a bit difficult to understand (especially for "days"). I wonder if there's chance to go back to the team and revisit the design before we decide on the dev approach?

cc @sherakama @rebeccahongsf

Comment (yvonnetangsu) [reply package.json:8]: Would rather us not have to install emotion just for this one thing 😄

Comment (forrestglick) [reply package.json:8]: Thanks, all, for your thoughts on how best to implement this countdown timer. A similar timer will (probably) be included in the emails going out to donors using this service: https://countdownmail.com/ The team wanted to visually match the look of the countdown timer, hence the "rings" around the numbers. All that being said, I am happy to go back to the team and propose removing the rings to make the implementation cleaner and more efficient.

Comment (forrestglick) [reply package.json:8]: Ok, I just spoke with the project lead, and she has a strong desire to keep the rings as part of the display. She feels it creates an added urgency to take action before the time has run out and provides a similar visual to what users have seen in the email driving traffic to the page. As a fallback to writing our own version of the timer, we could also embed a third-party countdown timer – assuming we could find one that meets our needs.

Comment (yvonnetangsu) [reply package.json:8]: @forrestglick @wmui51 Ahh I see - totally understand trying to match the look of the email timer. As someone who hasn't seen the email, my initial impression was that the rings add to the cognitive load, and I was more drawn to thinking about the ring fragments instead of on the time left instead (which probably isn't what you all want 😄 ). Would be awesome if we could change the design - if not, I think css vars might be worth an exploration @wmui51 👍🏼

Comment (yvonnetangsu) [reply package.json:8]: @forrestglick Sorry I replied before seeing your latest message 😬

Comment (forrestglick) [reply package.json:8]: @yvonnetangsu Yes, I agree the rings add to the cognitive load, and I also want to defer to the project lead and designer. Do know that I've explained a bit about the complexities of the implementation, so they are at least aware of what is involved. I think it would be great if Wilbert explored alternatives to loading the emotion package. If we use it for this one page, would that JS package get loaded on any call to the site? Any means of keeping the code clean/efficient sounds good to me!

Comment (yvonnetangsu) [reply package.json:8]: @forrestglick Thanks for bringing that up 😄 I believe we could try different approach before installing the package. @wmui51 I've had success in the past by creating an empty `div` to replace a pseudo element. Perhaps give that a try?

Comment (forrestglick) [reply package.json:8]: @wmui51 Yes, I was thinking of a service that basically loaded as an iframe-type embed. That way, it would only load when that specific page was rendered. I leave the explorations to you all, as I'm clearly out of my depth.

Comment (yvonnetangsu) [reply package.json:8]: I'm fine with iframe embed as well if we could find one that matches the look and feel, provided that it's accessible 😄 @forrestglick @wmui51

Comment (forrestglick) [reply package.json:8]: I brief search online surfaced this option: https://logwork.com/countdown-1or7

Comment (yvonnetangsu) [reply package.json:8]: That looks really promising @forrestglick 😄  ! Any chance the text (days, hours etc) can be a bit bigger. They're quite tiny on mobile.

Comment (forrestglick) [reply package.json:8]: @yvonnetangsu I do not see a way to control for text size – only colors can be adjusted.

Comment (sherakama) [reply package.json:8]: @wmui51 and @yvonnetangsu, 

This is a SCSS and JavaScript based product. Why not just programmatically generate all the styles using a loop and rotate through the class names?

Comment (yvonnetangsu) [reply package.json:8]: @sherakama That works too - we were using a loop to generate TW-liked classes for background colors etc in this repo already, so open to using a loop t generate rotation classes as well. Though I'm ok with limited use of inline styles as well - since we do use that with dynamic background images and transforms.

Comment (sherakama) [reply package.json]: Thanks @wmui51 

I appreciate the refactor. I am fine with this approach.

Comment (yvonnetangsu) [reply package.json]: @wmui51 Also fine with this approach - thank you 👍🏼

**Comment by yvonnetangsu** [comment]
Hi @wmui51 ! I heard that you would like to turn this into a Storyblok component. Here is a Storyblok tutorial to get your started:
https://www.storyblok.com/tp/create-custom-components-in-storyblok-and-gatsby

I'd suggest taking a look at our existing components and compare to how they're set up in the Storyblok backend in the `DEV OOD Giving` Storyblok space. I'm suspecting that the count down clock will have fairly straightforward field types - strings and numbers perhaps? Some of the card components, eg, this one, might be a good place to start:
https://github.com/SU-SWS/ood_giving_site/blob/dev/src/components/cards/oodSupportCard.js

cc @forrestglick

**Comment by forrestglick** [comment]
@yvonnetangsu Adding configurable options to the component is not a requirement for this initial release. The team is hoping to begin testing the email-to-landing page flows early next week, so I'd like to prioritize getting this pushed to Dev, if possible. Please let Wilbert know if there are any blockers. Thank you!

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu Adding configurable options to the component is not a requirement for this initial release. The team is hoping to begin testing the email-to-landing page flows early next week, so I'd like to prioritize getting this pushed to Dev, if possible. Please let Wilbert know if there are any blockers. Thank you!

Sounds like a plan @forrestglick! I'll review later today.

**Comment by yvonnetangsu** [comment]
@wmui51 The deploy preview is linked to the DEV Storyblok environment, could you publish a test page on DEV storyblok so we can get a preview? You said previously it's on 
https://deploy-preview-379--adapt-giving.netlify.app/endowed-positions/graduate-school-of-business/
but looks like it's not there anymore?

**Comment by yvonnetangsu** [comment]
And, from the screenshot, there doesn't seem to be spacing between the circles - is that intentional 🤔 ? @wmui51 @forrestglick

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu i removed it from the test page in the finalized pr because it isn't supposed to live on that page. my understanding is there is no giving tuesday page setup on dev. i am trying to get this merged into dev so the component can become available on the dev storyblok to drop in like this pr: #260.
> 
> I can put it back on a page for inspection, but we would need to remove it before merging, let me know how you want to proceed. As for the styles, I had it w/o spacing ever since the request to remove the `:` between rings and last i checked it passed design review, but can add spacing if requested.
> 
> cc @forrestglick

@wmui51 Yes, that'd be fine - please feel free to put it on any page like you did before, and then we can remove it before merging 👍🏼

**Comment by yvonnetangsu** [comment]
As for the spacing - it is up to your team of course 😄 I was looking previous example that @forrestglick shared it does have some spacing. If your team is ok with no spacing, I'm ok with that. 
https://logwork.com/countdown-1or7

**Comment by forrestglick** [comment]
I agree some spacing between the countdown items would be preferred. Good catch, Yvonne! And thanks for the guidance on making this into a Storyblok component – I know your plate is totally full.

**Comment by yvonnetangsu** [comment]
@forrestglick Happy to circle back on the Storyblok component part afterwards. If it's only ever going to be on one page, then I don't think we need the picker approach even 😄

**Comment by yvonnetangsu** [comment]
@wmui51 Thanks for the preview link - a few comments 😄 

- The font size for XS (eg, 360px) for the "hours" etc text is super small (7px). Could we make them bigger? Perhaps instead of uppercase (uppercase text is wider and harder to read), change the text to regular case, but 10px instead of 7px?
- Seeing some tiny glitches when you resize the browser (the little horizontal lines)
Screenshot 2023-10-26 at 4 59 49 PM
- I imagine that this clock will come with a heading or some text that describe what it does - otherwise you need need a screen reader only heading. I suggest going to the SODA office hour to get a recommendation for whether you need to also announce the changes in seconds via aria live region or something. Sign up for office hour here:
https://stanford.link/soda-help
- Spacing looks good - thanks!

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu I can't reproduce the resize glitch I was wondering what browser/platform you are using? I tried Mac safari, windows ff/chrome and was not getting that.
> 
> I can set the font, I'm setting the min font size at .4rem, but can change that to 1rem. Also @forrestglick would you be ok not using uppercase.
> 
> As for the accessibility parts, @forrestglick do you know if there will be a heading? Also I'm not sure on the timing for everything but can this wait until post-SODA office hours to deploy or is that something we can iterate on?

@wmui51 I'm using Chrome on Mac. And when would you like to go live - I understand that you'd like to get this into dev as soon as possible which is ok since that's not public @forrestglick

**Comment by yvonnetangsu** [comment]
Ahh I went back and read your comment - so you want to test early next week for emails 👍🏼 @forrestglick I'm ok with merging this into dev first and iterate. Let me review the code some more.

**Comment by yvonnetangsu** [comment]
@wmui51 I'm out tomorrow, so I'll wait till @forrestglick has a chance to comment on the font issue. I'll check back on Monday for a review and help get this into dev as soon as we can.

**Comment by forrestglick** [comment]
Good morning, @yvonnetangsu.  Wilbert and I are hoping you can review/approve this PR to get it into Dev for further testing. Thanks!

**Review by yvonnetangsu — APPROVED** [review]
@forrestglick @wmui51 I understand that you need this ASAP for merging into dev for further testing so approving. I'll pull it down later to test it locally. @wmui51 Please feel free to push up a commit to remove the code from the endowed positions page, and add the block to another page of your choice for testing.

**Comment by yvonnetangsu** [comment]
@wmui51 @forrestglick Would be good to follow up with SODA (office hour probably easiest) for a quick check after you set up a page on DEV for this timer to live.

File: src/components/giving-tuesday/Countdown.js
Comment (yvonnetangsu) [inline src/components/giving-tuesday/Countdown.js:7]: @wmui51 Have you tested with people at different time zones? This looks like it should work as expected, but probably won't hurt to double check. I'm going to merge this in so you all can do the testing.

**Review by yvonnetangsu — APPROVED** [review]
Approve and merging into dev 👍🏼

**Comment by yvonnetangsu** [comment]
> i did not test with a changed timezone, but did encounter some timezone issues early on and ended up passing in the timeZone param.

Thanks - it is merged now so please feel free to test on DEV. To be safe, I'd suggest send it to users in different time zone for testing and see if it's behaving like you expected.

**Comment by forrestglick** [comment]
@wmui51 I ran a test using the page: https://dev--adapt-giving.netlify.app/giving-tuesday-test/ by changing my local timezone. The countdown clock changed, so we will need to re-factor to keep it a consistent count for all users. We will also need to accommodate 36+ hours of counting, as the campaign will launch "early" for a select group of donors. Campaign should end at midnight when the date changes from the 28th to 29th.

**Comment by forrestglick** [comment]
@wmui51 correct, the giving tuesday campaign ends for **everyone** at midnight of Nov 28 (as the clock turns to the 29th) and will launch mid-day on the 27th for a select group of donors. Therefore, the countdown should **not** take the local time into account, but instead display as it does for Pacific time for all users. I'm ok with displaying the days. In fact, after I wrote you, I thought better of my message above and agree displaying the days will be best. Thanks!

**Comment by yvonnetangsu** [comment]
> @wmui51 correct, the giving tuesday campaign ends for **everyone** at midnight of Nov 28 (as the clock turns to the 29th) and will launch mid-day on the 27th for a select group of donors. Therefore, the countdown should **not** take the local time into account, but instead display as it does for Pacific time for all users. I'm ok with displaying the days. In fact, after I wrote you, I thought better of my message above and agree displaying the days will be best. Thanks!

Just an additional thought, if we're displaying the day circle, perhaps it could be on its own line, so on mobile (XS breakpoint), we have enough space to have things at their current font size.

**Comment by forrestglick** [comment]
Good suggestion, Yvonne. And once the days are "zero", it would be great if the days circle did not display.

**Comment by forrestglick** [comment]
Alexis is thinking the entire block will be centered. Same for the days (above). I am able to center the section title, but not the timer, currently.

**Pull request merged**
